Rework publication fetching

MatthiasZepper opened the pull request:
## Motivation

Over the past months, we encountered recurring issues with our website being fully unresponsive or slow when loading. 

After some investigations, we could identify the publication list on the landing page as the culprit. When fetching publications from [`publications.scilifelab.se`](https://publications.scilifelab.se), it would need to process a huge JSON file with thousands of publications, which sometimes overwhelmed the web server. 

Since it is currently not possible to request a lesser number of items from the SciLifeLab Publications API, I added a separate caching layer in the process that gives us control over the number of publications received by our website. A [GitHub Action that runs weekly](https://github.com/NationalGenomicsInfrastructure/ngisweden.se-publications/blob/main/.github/workflows/cronjob.yml) and [caches 50 publications](https://github.com/NationalGenomicsInfrastructure/ngisweden.se-publications/tree/main/cache) every time to be displayed on our website.

## Content of this pull request

This pull request now introduces a new shortcode `ngisweden_publications_gh`, that shares all functionality with the old ngisweden_publications, but uses our GitHub cache instead of the SciLifeLab publications API. This results in significantly faster loading times and should mitigate the occasional unresponsiveness of the website. 

## Testing

Currently, we do not have a fully-fledged testing environment for our website available. However, for testing this particular feature, it is also not required.

To test:

1. Clone the branch underlying this pull request locally and navigate to the directory.
2. Run `docker compose up`. (_requires Docker to be installed_)
3. Open `localhost:8000` in the browser of your choice.
4. Follow through the WordPress installation dialogue and create a blank instance of the website.
5. [Log into the system](http://localhost:8000/wp-login.php) with your newly created credentials.
6. On the [Plugins page](http://localhost:8000/wp-admin/plugins.php), activate the `NGI Custom Content` plugin.
7. On the [Themes page](http://localhost:8000/wp-admin/themes.php), activate the `NGI Sweden Theme`. 
8. Create a new blog post, news item or page (or edit one of the samples). In the visual editor, create a `Shortcode` block. You can just type `/Shortcode` to do this. 
9. As content for the short code, enter `[ngisweden_publications_gh num=7 collabs=1]` to obtain the block as rendered on [NGI landing page](https://ngisweden.scilifelab.se)
10. Save and publish the new content. Experiment with it and report warnings or failures.

MatthiasZepper committed:
Update docker-compose

MatthiasZepper committed:
Fix strtoupper() and strtolower() on potentially null values from the author data.

MatthiasZepper committed:
Create a new shortcode that fetches the publications from the GitHub cache instead of SciLifeLab publications.

MatthiasZepper committed:
Better maintain a separate cache for the GH publication shortcode.

MatthiasZepper committed:
Add checks for empty values in JSON to prevent warnings:

Added checks for all journal fields:
 - journal['title']
 - journal['volume']
 - journal['issue']
 - journal['issn']

Added checks for publication metadata:
 - title
 - published
 - abstract
 - pmid
 - doi
 - iuid

Added checks for flags and links:
 - is_collab
 - is_tech_dev
 - links['display']['href']

alneberg reviewed (approved):
Great instructions to test it! And everything works as intended!

MatthiasZepper committed:
Make the publication fetching more resilient against errors:

- Added a timeout of 10 seconds to prevent the request from hanging indefinitely
- Added a proper user agent string to identify the requests
- Implemented a fallback mechanism that will:
	- Try to use the local cache if GitHub is unreachable
	- Only use the local cache if it's less than 30 days old

- Added initialization of $pubs_data['publications'] as an empty array at the start if it's not set or not an array
- Replaced @count() with empty() for more reliable array checking
- Added proper null checks before processing publications
- Wrapped the publication processing code in an if (!empty($pubs_data['publications'])) check
- Added an additional safety check before proceeding with the main logic
- Made sure the randomization only happens if we have publications
- Added proper type checking using isset() and is_array()

MatthiasZepper commented:
Per direct conversation with Johannes, I have used AI to code a more resilient error handling strategy that accounts for a deleted cache and a failed retrieval from GitHub. I have tested this by manually taking my local machine offline and deleting the cache. In that case, now an error is printed to the screen, but the publication fetching module does not crash.

alneberg commented:
I can confirm the new changes work great! 👍

MatthiasZepper committed:
README.md
